sol/main.c

```diff
@@ -0,0 +1,513 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "gatekeeper_gk.h"
+#include "gatekeeper_launch.h"
+#include "gatekeeper_sol.h"
+
+#define DEFAULT_TB_SIZE (5000)
+
+/*
+ * Solicitor configuration allocated by alloc_sol_conf(), needed
+ * by gk_solicitor_enqueue() since the GK instances do not have
+ * any visibility into the Solicitor.
+ */
+static struct sol_config *sol_conf;
+
+static void
+insert_new_priority_req(struct req_queue *req_queue, struct priority_req *pr,
+	uint8_t priority)
+{
+	uint8_t next;
+
+	req_queue->priorities[priority] = pr;
+
+	/* This is the first packet in the queue. */
+	if (req_queue->len == 0) {
+		req_queue->head = pr;
+		req_queue->highest_priority = priority;
+		req_queue->lowest_priority = priority;
+		return;
+	}
+
+	/* Insert at head of queue. */
+	if (priority > req_queue->highest_priority) {
```

Comment:
A higher value of `priority` represents a higher priority, so if the priority of the packet we're inserting is larger than the highest priority that's currently in the queue (larger than `req_queue->highest_priority`), then we insert it at the head of the queue.